lib/index.js.flow

```diff
@@ -0,0 +1,188 @@
+ * seems can not resolve a Thenable's value type directly
+ * so `Query extends Promise` to make thing temporarily work.
+ * like this:
+ * const q = client.query('select * from some');
+ * q.on('row',cb); // Event
+ * const result = await q; // or await
+ *
+ * ToDo: should find a better way.
+*/
+declare class Query extends Promise<ResultSet> {
+  then<U>( onFulfill?: (value: ResultSet) => Promise<U> | U,
+    onReject?: (error: PG_ERROR) => Promise<U> | U
+  ): Promise<U>;
+  // Because then and catch return a Promise,
+  // .then.catch will lose catch's type information PG_ERROR.
+  catch<U>( onReject?: (error: PG_ERROR) => ?Promise<U> | U ): Promise<U>;
+
+  on :
+  ((event: 'row', listener: (row: Row, result: ResultBuilder) => void) => EventEmitter )&
+  ((event: 'end', listener: (result: ResultBuilder) => void) => EventEmitter )&
+  ((event: 'error', listener: (err: PG_ERROR) => void) => EventEmitter );
+//  ((event: string, listener: Function) => EventEmitter );
+}
+
+/*
+ * lib/client.js
+ * Note: if not extends from EventEmitter.
+ * on() can return this. Flow's EventEmitter force return a EventEmitter in on()
+*/
+declare class Client {
+  constructor(config?: string | ClientConfig): void;
+  connect(callback?: ClientConnectCallback):void;
+  end(): void;
+
+  query:
+  ( (query: QueryConfig|string, callback?: QueryCallback) => Query ) &
+  ( (text: string, values: Array<any>, callback?: QueryCallback) => Query );
+
+  on:
+  ((event: 'drain', listener: () => void) => this )&
+  ((event: 'error', listener: (err: PG_ERROR) => void) => this )&
+  ((event: 'notification', listener: (message: any) => void) => this )&
+  ((event: 'notice', listener: (message: any) => void) => this )&
+  ((event: 'end', listener: () => void) => this );
+//  ((event: string, listener: Function) => this);
+}
+
+/*
+ * require('pg-types')
+*/
+type TypeParserText = (value: string) => any;
+type TypeParserBinary = (value: Buffer) => any;
+type Types = {
+  getTypeParser:
+//    ((oid: number, format: 'text') => TypeParserText )&
+    ((oid: number, format?: 'text') => TypeParserText )&
+    ((oid: number, format: 'binary') => TypeParserBinary );
+
+  setTypeParser:
+    ((oid: number, format?: 'text', parseFn: TypeParserText) => void )&
+  // ( (oid: number, parseFn: TypeParserText) => void )&
+    ((oid: number, format: 'binary', parseFn: TypeParserBinary) => void),
+}
+
+/*
+ * lib/index.js ( class PG)
+*/
+declare class PG extends EventEmitter {
+  types: Types;
+  Client: Class<Client>;
+  Pool: Class<Pool>;
+  Connection: any; //Connection is used internally by the Client.
+  constructor(client: Client): void;
+  native: { // native binding, have the same capability like PG
+    types: Types;
+    Client: Class<Client>;
+    Pool: Class<Pool>;
+    Connection: any;
+  };
+// The end(),connect(),cancel() in PG is abandoned ?
+}
+
+// module export, keep same structure with index.js
+declare var pg:PG ;
```

Comment:
marked, will amend them tomorrow, when the review finished, will also delete some comments only used for review purpose.
Sleeping, see you tomorrow. Thanks for review.